### PR TITLE
[SPARK-50055][SQL] Add TryMakeInterval alternative

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Addition of two new expression try_make_interval.

### Why are the changes needed?
This is a split PR from https://github.com/apache/spark/pull/48499 so that we divide the reasonings for PRs.

### Does this PR introduce _any_ user-facing change?
Yes, new expressions added.

### How was this patch tested?
Tests added.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: python/pyspark/sql/connect/functions/builtin.py [hunk 1]
```diff
@@ -617,6 +617,13 @@ def ceiling(col: "ColumnOrName", scale: Optional[Union[Column, int]] = None) ->
 ceiling.__doc__ = pysparkfuncs.ceiling.__doc__
 
 
+def try_conv(col: "ColumnOrName", fromBase: int, toBase: int) -> Column:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/functions/builtin.py:620]: Let's add a test at `test_functions.py` so the tests can be shared

**Review by MaxGekk — COMMENTED** [review]
Don't think it is good idea to mix two new expressions in one PR especially when they are unrelated. I believe it is better to open separate PRs per expressions.

**Review by MaxGekk — COMMENTED** [review]
@mihailom-db Could you resolve conflicts and fix the test, it seems it is related to your changes:
```
ERROR [0.217s]: test_try_make_interval (pyspark.sql.tests.test_functions.FunctionsTests.test_try_make_interval)
```

File: python/pyspark/sql/connect/functions/builtin.py [hunk 2]
```diff
@@ -3712,6 +3712,31 @@ def make_dt_interval(
...
 
 
+def try_make_interval(
+        years: Optional["ColumnOrName"] = None,
+        months: Optional["ColumnOrName"] = None,
+        weeks: Optional["ColumnOrName"] = None,
+        days: Optional["ColumnOrName"] = None,
+        hours: Optional["ColumnOrName"] = None,
+        mins: Optional["ColumnOrName"] = None,
+        secs: Optional["ColumnOrName"] = None,
```
Comment (MaxGekk) [inline python/pyspark/sql/connect/functions/builtin.py:3722]: Please, fix indentations here.

File: python/pyspark/sql/functions/builtin.py [hunk 3]
```diff
@@ -21152,6 +21152,162 @@ def make_dt_interval(
...
 
+@_try_remote_functions
+def try_make_interval(
+        years: Optional["ColumnOrName"] = None,
+        months: Optional["ColumnOrName"] = None,
+        weeks: Optional["ColumnOrName"] = None,
+        days: Optional["ColumnOrName"] = None,
+        hours: Optional["ColumnOrName"] = None,
+        mins: Optional["ColumnOrName"] = None,
+        secs: Optional["ColumnOrName"] = None,
```
Comment (MaxGekk) [inline python/pyspark/sql/functions/builtin.py:21163]: fix indentations

File: python/pyspark/sql/functions/builtin.py [hunk 4]
```diff
@@ -21152,6 +21152,162 @@ def make_dt_interval(
...
+    >>> df = spark.createDataFrame([[100, 11, 1, 1, 12, 30, 01.001001]],
+    ...     ["year", "month", "week", "day", "hour", "min", "sec"])
+    >>> df.select(sf.try_make_interval(df.year)).show(truncate=False)
+    +-----------------------------------------+
+    |try_make_interval(year, 0, 0, 0, 0, 0, 0)|
+    +-----------------------------------------+
+    |100 years                                |
+    +-----------------------------------------+
+
+    Example 8: Try make interval.
```
Comment (MaxGekk) [inline python/pyspark/sql/functions/builtin.py:21286]: Could you add an example which shows the difference between try and non-try versions.

File: sql/api/src/main/scala/org/apache/spark/sql/functions.scala [hunk 5]
```diff
@@ -8073,6 +8161,16 @@ object functions {
...
     Column.fn("make_interval", years)
 
+  /**
+   * This is a special version of `make_interval` that performs the same operation, but returns a
+   * NULL value instead of raising an error if interval cannot be created.
+   *
+   * @group url_funcs
+   * @since 4.0.0
+   */
+  def try_make_interval(): Column =
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/functions.scala:8171]: Why do you need this one? In which cases does `make_interval()` fail?

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/api/src/main/scala/org/apache/spark/sql/functions.scala (hunk 5 above)
Comment (mihailom-db) [reply sql/api/src/main/scala/org/apache/spark/sql/functions.scala:8171]: You are right, we cannot get an error here, as this is the same as makeinterval(0) which is not an overflow. Will update the pr.

**Review by MaxGekk — APPROVED** [review]

**Review by MaxGekk — COMMENTED** [review]
Waiting for CI.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by mihailom-db** [comment]
@MaxGekk @HyukjinKwon Do we want to not support try_make_interval in pyspark? This PR seems to block it https://github.com/apache/spark/pull/46975. I believe we should update that make_interval is not supported as well, as CalendarIntervalType is not supported.

**Comment by mihailom-db** [comment]
Should I just then remove it from python API? As we cannot really call it there.

**Comment by HyukjinKwon** [comment]
hm, we can call them, no? It's just that we don't support `df.collect()` on `CalendarIntervalType` instances.

**Comment by HyukjinKwon** [comment]
e.g., you can call `df.select(make_interval(...)).show()`

**Comment by mihailom-db** [comment]
Aha, ok, so I should only change the test to use some other type of collection of data that is not `collect()` and then verify the results.

**Comment by HyukjinKwon** [comment]
Yeah, i think so.

**Comment by mihailom-db** [comment]
Ready for merge now @MaxGekk

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @mihailom-db and @HyukjinKwon for review.

**Pull request closed without merging**